File: sorting_index/test_sorting_sorting_addnoise_time.py
import csv
import time

import numpy as np
from scipy.io import loadmat

fileName = "../data/data_static_indoor_1.mat"
rawData = loadmat(fileName)

CSIa1Orig = rawData['A'][:, 0]
CSIb1Orig = rawData['A'][:, 1]
dataLen = len(CSIa1Orig)

segLen = 7
keyLen = 256 * segLen

times = 0
overhead = 0

for staInd in range(0, dataLen, keyLen):
    endInd = staInd + keyLen
    print("range:", staInd, endInd)
    if endInd >= len(CSIa1Orig):
        break
    times += 1

    CSIa1Orig = rawData['A'][:, 0]
    CSIb1Orig = rawData['A'][:, 1]
    CSIn1Orig = np.random.normal(loc=1, scale=2, size=dataLen)

    seed = np.random.randint(10000)
    np.random.seed(seed)
    noiseOrig = np.random.uniform(0, np.std(CSIa1Orig) * 3, size=dataLen)
    CSIa1Orig = (CSIa1Orig - np.mean(CSIa1Orig)) + noiseOrig
    CSIb1Orig = (CSIb1Orig - np.mean(CSIb1Orig)) + noiseOrig
    CSIn1Orig = (CSIn1Orig - np.mean(CSIn1Orig)) + noiseOrig

    tmpCSIa1 = CSIa1Orig[range(staInd, endInd, 1)]
    tmpCSIb1 = CSIb1Orig[range(staInd, endInd, 1)]

    tmpCSIa1Ind = np.array(tmpCSIa1).argsort().argsort()

    start = time.time()
    tmpCSIb1Ind = np.array(tmpCSIb1).argsort().argsort()
    minEpiIndClosenessLsb = np.zeros(int(keyLen / segLen), dtype=int)

    for i in range(int(keyLen / segLen)):
        epiInda1 = tmpCSIa1Ind[i * segLen:(i + 1) * segLen]

        epiIndClosenessLsb = np.zeros(int(keyLen / segLen))

        for j in range(int(keyLen / segLen)):
            epiIndb1 = tmpCSIb1Ind[j * segLen: (j + 1) * segLen]

            epiIndClosenessLsb[j] = sum(abs(epiIndb1 - np.array(epiInda1)))

        minEpiIndClosenessLsb[i] = np.argmin(epiIndClosenessLsb)
    print(minEpiIndClosenessLsb)

    end = time.time()
    overhead += end - start
    print("time:", end - start)

print(overhead / times)

File: sorting_index/plot/no additive nosie/test_sorting_sorting_addnoise_time.m
segLen=7;
keyLen=256*segLen;
times=0;
overhead=0;

load('./data_static_indoor_1.mat');
CSIa10rig=A(:,1);
CSIb10rig=A(:,2);
dataLen=length(CSIa10rig);

% poolSize=4;
% parpool("local",poolSize);

for staInd=1:keyLen:dataLen
    endInd=staInd+keyLen;
    fprintf("range: %d %d",staInd,endInd);
    if endInd > length(CSIa10rig)
        break
    end
    times=times+1;
    CSIn10rig=normrnd(1,2,[dataLen,1]);
    seed=randi(10000);
    rng(seed);
    noise0rig=rand(dataLen,1).*3.*std(CSIa10rig);
    CSIa10rig=(CSIa10rig-mean(CSIa10rig))+noise0rig;
    CSIb10rig=(CSIb10rig-mean(CSIb10rig))+noise0rig;
    CSIn10rig=(CSIn10rig-mean(CSIn10rig))+noise0rig;

    tmpCSIa1=CSIa10rig(staInd:endInd);
    tmpCSIb1=CSIb10rig(staInd:endInd);
    [out,idx]=sort(tmpCSIa1);
    [out,tmpCSIa1Ind]=sort(idx);
%     tic;
    t1 = clock;
    [out,idx]=sort(tmpCSIb1);
    [out,tmpCSIb1Ind]=sort(idx);

    minEpiIndClosenessLsb=zeros(keyLen/segLen);
    loop_time=keyLen/segLen;
    for i=0:loop_time-1
        epiInda1=tmpCSIa1Ind(i*segLen+1:(i+1)*segLen);
        epiIndClosenessLsb=zeros(loop_time,1);
        for j=0:loop_time-1
            epiIndb1=tmpCSIb1Ind(j*segLen+1:(j+1)*segLen);
            epiIndClosenessLsb(j+1)=sum(abs(epiIndb1-epiInda1));
        end
        [~,argmin]=min(epiIndClosenessLsb);
        minEpiIndClosenessLsb(i+1)=argmin;
    end
%     toc
    t2 = clock;
    overhead = overhead + etime(t2,t1);
end

overhead = overhead / times;